server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java

```diff
@@ -2091,6 +2092,45 @@ protected long transportDelayMillis(String actionName) {
         }
     }
 
+    public void testGetClusterFormationStateDoesNotBlockOnMutex() {
+        // Pick any node in the cluster (all have a Coordinator)
+        try (Cluster cluster = new Cluster(1)) {
+            cluster.stabilise();
+            final Coordinator coordinator = cluster.getAnyNode().coordinator;
+
+            // Latches for coordination
+            final var mutexAcquired = new CountDownLatch(1);
+            final var release = new CountDownLatch(1);
+
+            // Start a thread that acquires the mutex and blocks
+            Thread t = new Thread(() -> {
+                synchronized (coordinator.mutex) {
+                    mutexAcquired.countDown();
+                    try {
+                        release.await();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    }
```

Comment:
Definitely shouldn't be waiting forever here, and should fail on an interrupt:

```suggestion
                    mutexAcquired.countDown();
                    safeAwait(release);
```

But also this could just be

```suggestion
                    safeAwait(barrier);
                    safeAwait(barrier);
```